blog/en-us/spring-boot-dubbo-start-stop-analysis.md

```diff
@@ -0,0 +1,223 @@
+
+In the await method, the current thread checks the variable `stopAwait` every 10 seconds in a while loop. It is a `volatile` variable that is used to ensure that the current thread can see the change immediately after being modified by another thread. If there is no change, it will always be in the while loop. This is why the thread does not exit, which is why the entire Spring-boot application doesn't exit.
+
+Because Spring-boot application enable 2 port that 8080 and 8081(management port) at the same time. Two Tomcat are startup actually, so there are two threads that `container-0` and `container-1`.
+
+Then, let's see how this Spring-boot application exits.
+
+### The analysis of DubboConsumer exit
+
+As mentioned in the previous description, there is a thread that checks the variable `stopAwait` continuously. Then we can think that should be a thread to modify `stopAwait` at Stop, it can break the while loop. So, who is modifying this variable?
+
+By analyzing the source code, you can see that there is only one way to modify `stopAwait`, in `org.apache.catalina.core.StandardServer#stopAwait`, we add a breakpoint here to see who is calling.
+
+> Note that when we add a breakpoint in Intellij IDEA's Debug mode, we need to use `kill -s INT $PID` or `kill -s TERM $PID` on command line to trigger the breakpoint. The breakpoint will not be trigger if you just click the stop button on the IDE.
+
+You can see the method is called by a thread called `Thread-3`:
+
+```java
+stopAwait:390, StandardServer (org.apache.catalina.core)
+stopInternal:819, StandardServer (org.apache.catalina.core)
+stop:226, LifecycleBase (org.apache.catalina.util)
+stop:377, Tomcat (org.apache.catalina.startup)
+stopTomcat:241, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+stop:295, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+stopAndReleaseEmbeddedServletContainer:306, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+onClose:155, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+doClose:1014, AbstractApplicationContext (org.springframework.context.support)
+run:929, AbstractApplicationContext$2 (org.springframework.context.support)
+```
+
+Through source code analysis, it was executed by Spring's registered `ShutdownHook`.
+
+```java
+    @Override
+    public void registerShutdownHook() {
+        if (this.shutdownHook == null) {
+            // No shutdown hook registered yet.
+            this.shutdownHook = new Thread() {
+                @Override
+                public void run() {
+                    synchronized (startupShutdownMonitor) {
+                        doClose();
+                    }
+                }
+            };
+            Runtime.getRuntime().addShutdownHook(this.shutdownHook);
+        }
+    }
+```
+
+We found that ShutdownHook will execute in the following two cases by referring the Java API documentation[2].
```

Comment:
By reffering the Java API documentation[2], we found that .... will be executed under the following two cases